### DISCUSSION TITLE
GMMconv only learn to aggregate features from the node itself (self-loop)

### DISCUSSION THREAD

**Discussion by Icecream-blue-sky** [pyg-team/pytorch_geometric#3903] [Q&A]
When I was using GMMconv to learn spatial relation of objects in object detection task, I found that the GMMconv only learn to aggregate features from the node itself. The edge between object i and object i is much larger than that between object i and other object. I use polar pseduo coordinates based on objects' centers, which is like below:
![image](https://user-images.githubusercontent.com/26771037/150458383-e5af8b5c-cd6b-47fd-91db-f8608a3673d6.png)
I send them to GMMconv after normalizing the pseduo coordinates. In the graph, the node feature is the object feature. The graph have self-loop before sending to GMMconv.
```python
edge_spatial = edge_spatial/torch.tensor([1333, 2*math.pi]).unsqueeze(0).cuda() # 1333 is the max width or height of the image.
```
I set the aggr = 'mean',  root-weight = True (which may be redundant? Is it the cause?), and kernel_size=16. 
What is the cause of this problem? Should I do special handling for self-loop edge? Since its polar pseduo coordinate is (0,0).

**Comment by rusty1s** [comment]
How do you determine that `GMMConv` ignores neighboring node information? As you correctly pointed out, there exists two ways to incorporate "self-loops":
1. `root_weight=True` will learn a separate transformation/skip-connection for central node features
2. As part of a self-loop: The layer can only separate neighbor and central information based on pseudo coordinates.

IMO, dropping the self-loop and apply `root_weight=True` is probably the better option.

**Reply by Icecream-blue-sky** [reply]
> How do you determine that `GMMConv` ignores neighboring node information? As you correctly pointed out, there exists two ways to incorporate "self-loops":
> 
> 1. `root_weight=True` will learn a separate transformation/skip-connection for central node features
> 2. As part of a self-loop: The layer can only separate neighbor and central information based on pseudo coordinates.
> 
> IMO, dropping the self-loop and apply `root_weight=True` is probably the better option.

Maybe you are right, I'll try. Thanks.
